include/crm/msg_xml.h

```diff
@@ -219,13 +219,19 @@ extern "C" {
 
 #  define XML_CIB_TAG_RSC_TEMPLATE	"template"
 
-#  define XML_RSC_ATTR_CLEAR_OP         "clear_failure_op"
-#  define XML_RSC_ATTR_CLEAR_INTERVAL   "clear_failure_interval"
 #  define XML_RSC_ATTR_REMOTE_RA_ADDR   "addr"
 #  define XML_RSC_ATTR_REMOTE_RA_SERVER "server"
 #  define XML_RSC_ATTR_REMOTE_RA_PORT   "port"
 #  define XML_RSC_ATTR_CRITICAL         "critical"
 
+/* @TODO These are op attrs that are never set anywhere. Evaluate whether they
+ * (and the code added with them in 5ddfc16f) can be dropped.
+ *
+ * If kept, they should be renamed to reflect that they're op attrs.
```

Comment:
Now that I know these are for T382, only renaming them remains to be done.

That raises a question that I've been pondering: do we want to stick with our current naming approach? For reference, that is:
* `PCMK_XE_*`
* `PCMK_XA_*`
* `PCMK__META_*`

There are two subquestions here:
1. Do we want to stick with public names? Obviously I've gone with internal names for `PCMK__META_*` in this PR, because there was precedent prior to (public) `PCMK_META_ENABLED`. If so, what should be public and what should be internal? It might be reasonable for element names -- perhaps only CIB element names -- to be public. It's harder to envision public use cases for XML attributes, and even harder to envision public use cases for meta-attribute names.
2. Are these prefixes sufficiently descriptive and accurate?

To illustrate a couple of problems with #2:
* Some things (e.g., "enabled") can be used equally well as both XML attributes and meta-attribute names.
  * Used only as XML attribute for operations (not yet supported as a meta-attribute there, though it should be)
  * Used only as meta-attribute for alerts
* Operations are documented as accepting either element attributes or meta-attributes. This means the same constant should be both `PCMK_XA_*` and `PCMK__META_*`.
  * To a lesser extent this is true for resources as well, although I don't think it's documented. Resource element attributes (e.g., `provider` and `type`) get added to the meta-attribute hash.

You can also make an argument that more "semantic" naming is clearer, by allowing grouping of all resource attributes, all operation attributes, all alert attributes, etc. This is basically what the old naming scheme does. It's kind of nice to see be able to know immediately both:
* "this is the set of attributes that will be used with object type X", and
* "this particular attribute will definitely be used with object type X (so that if I'm doing something with object type X's options, I won't overlook this one)"

Semantic naming allows us to have `PCMK__ALERT_ATTR_PATH` and `PCMK__ALERT_ATTR_ENABLED` (or `PCMK__ALERT_META_ENABLED`) -- they're used in different positions (XML attribute vs. meta-attribute) but are both for alerts.

It also allows us to distinguish from a large list of meta-attributes, which ones are for which object types.

---

The planned grouping of option metadata into comprehensive arrays with object type indicated, addresses some of this but not all of it.

Reply:
My inclination is:
* Symbols should be public if the user may specify them somewhere (especially the CIB), internal if they're only used internal to Pacemaker (such as daemon IPC/CPG message XML)
* I usually favor something more descriptive than XE/XA (for XML elements/attributes), but we have so many XML-related APIs that anything longer gets unwieldy
* The same literal string may have more than one constant for different contexts. For example if something can be an XML element name in one context and an enumerated option value in another. For meta-attributes that can be specified as either XML attributes or nvpair names, it's borderline -- using just the meta constant would be fine since the XML attribute is still in a meta-attribute context, but having two constants would also be fine for consistency. I don't think we need to distingiush different meta-attribute contexts (such as resource vs alert).

I expect the main categories we would need are:
* XA = XML attribute name
* XE = XML element name
* META = meta-attribute
* OPT = cluster option
* VALUE = enumerated value of an XML attribute, name/value pair, meta-attribute, cluster option etc.

Reply:
Sounds good :)

> using just the meta constant would be fine since the XML attribute is still in a meta-attribute context, but having two constants would also be fine for consistency

"just the meta constant" is my general preference there. We're usually only going to check for the name once, rather than checking separately for an XML attribute or a meta-attribute. (There are some exceptions, like "enabled" for operations -- we check only for an XML attribute currently, but hopefully that changes in the future.)